src/raftstore/store/store.rs

```diff
@@ -2786,6 +2742,74 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
+
+    fn on_check_peer_stale_state_tick(&mut self, event_loop: &mut EventLoop<Self>) {
+        let mut leader_missing = 0;
+        for peer in &mut self.region_peers.values_mut() {
+            if peer.pending_remove {
+                continue;
+            }
+
+            if peer.is_applying_snapshot() || peer.has_pending_snapshot() {
+                continue;
+            }
+
+            // If this peer detects the leader is missing for a long long time,
+            // it should consider itself as a stale peer which is removed from
+            // the original cluster.
+            // This most likely happens in the following scenario:
+            // At first, there are three peer A, B, C in the cluster, and A is leader.
+            // Peer B gets down. And then A adds D, E, F into the cluster.
+            // Peer D becomes leader of the new cluster, and then removes peer A, B, C.
+            // After all these peer in and out, now the cluster has peer D, E, F.
+            // If peer B goes up at this moment, it still thinks it is one of the cluster
+            // and has peers A, C. However, it could not reach A, C since they are removed
+            // from the cluster or probably destroyed.
+            // Meantime, D, E, F would not reach B, since it's not in the cluster anymore.
+            // In this case, peer B would notice that the leader is missing for a long time,
+            // and it would check with pd to confirm whether it's still a member of the cluster.
+            // If not, it destroys itself as a stale peer which is removed out already.
+            match peer.check_stale_state() {
+                StaleState::Valid => (),
+                StaleState::LeaderMissing => {
+                    warn!(
+                        "{} leader missing longer than abnormal_leader_missing_duration {:?}",
+                        peer.tag, self.cfg.abnormal_leader_missing_duration.0,
+                    );
+                    leader_missing += 1;
+                }
+                StaleState::ToValidate => {
+                    // for peer B in case 1 above
+                    warn!(
+                        "{} leader missing longer than max_leader_missing_duration {:?}. \
+                         To check with pd whether it's still valid",
+                        peer.tag, self.cfg.max_leader_missing_duration.0,
+                    );
+                    let task = PdTask::ValidatePeer {
+                        peer: peer.peer.clone(),
+                        region: peer.region().clone(),
+                        merge_source: None,
+                    };
+                    if let Err(e) = self.pd_worker.schedule(task) {
+                        error!("{} failed to notify pd: {}", peer.tag, e)
+                    }
+                }
+            }
+        }
+        self.raft_metrics.leader_missing = leader_missing;
+
+        self.register_check_peer_stale_state_tick(event_loop);
```

Comment:
Why we need `register_check_peer_stale_state_tick ` on each tick?
What if leader_missing is zero?

Reply:
It needs to register the timeout again so that it will be invoked after the timeout.